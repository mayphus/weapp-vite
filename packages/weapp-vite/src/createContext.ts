import type { LoadConfigOptions } from './context'
import { getCompilerContext } from './context/getInstance'

export async function createCompilerContext(options?: Partial<LoadConfigOptions & { key?: string }>) {
  // 先初始化 ConfigService
  const ctx = getCompilerContext(options?.key)
  const { configService, scanService } = ctx
  await configService.load(options)
  // prefilght
  try {
    await scanService.loadAppEntry()
  }
  catch {
    // prefilght catch
  }

  return ctx
}
