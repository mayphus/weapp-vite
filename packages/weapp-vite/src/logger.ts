import logger from '@weapp-core/logger'

export default logger

export type LogType = 'error' | 'warn' | 'info'
export type LogLevel = LogType | 'silent'
