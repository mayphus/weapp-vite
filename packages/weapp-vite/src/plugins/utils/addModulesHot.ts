import type { PluginContext } from 'rollup'

export function addModulesHot(entriesSet: Set<string>, pluginContext: PluginContext) {
  for (const entry of entriesSet) {
    const moduleInfo = pluginContext.getModuleInfo(entry)

    if (moduleInfo) {
      const stack = [moduleInfo.id] // 用栈模拟递归
      const visitedModules = new Set<string>()

      while (stack.length > 0) {
        const id = stack.pop()

        if (id && !visitedModules.has(id)) {
          visitedModules.add(id)

          const info = pluginContext.getModuleInfo(id)

          if (info) {
            pluginContext.addWatchFile(info.id)
            // 将子依赖加入栈
            stack.push(...info.importedIds)
          }
        }
      }
    }
  }
}
