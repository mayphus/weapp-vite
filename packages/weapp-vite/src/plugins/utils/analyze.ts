import type { App, Component, Page } from '@weapp-core/schematics'
import { get } from '@weapp-core/shared'

export function analyzeAppJson(json: App) {
  const entries: string[] = []
  const pages = json.pages ?? []
  const components = Object.values(json.usingComponents ?? {})
  const subPackages = (
    [...json.subPackages ?? [], ...json.subpackages ?? []]?.reduce<string[]>(
      (acc, cur) => {
        acc.push(...(cur.pages ?? []).map(x => `${cur.root}/${x}`))
        return acc
      },
      [],
    )) ?? []

  if (get(json, 'tabBar.custom')) {
    entries.push('custom-tab-bar/index')
  }
  // 全局工具栏
  // https://developers.weixin.qq.com/miniprogram/dev/framework/runtime/skyline/appbar.html
  if (get(json, 'appBar')) {
    entries.push('app-bar/index')
  }
  entries.push(
    ...pages,
    ...components,
    ...subPackages,
  )
  return entries
}

export function analyzeCommonJson(json: Page | Component) {
  const entries: string[] = []
  const components = Object.values(json.usingComponents ?? {})
  entries.push(...components)
  return entries
}
