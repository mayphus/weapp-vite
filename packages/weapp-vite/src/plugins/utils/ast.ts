import type { Node } from 'estree'
// import type { ProgramNode } from 'rollup'
import { walk } from 'estree-walker'

export function collectRequireTokens(ast: Node) {
  const requireModules: {
    start: number
    end: number
    value: string
    async?: boolean
    // leadingComment: string
  }[] = []
  walk(ast, {
    enter(node) {
      if (node.type === 'CallExpression') {
        // TODO 可能要被废弃
        // https://developers.weixin.qq.com/miniprogram/dev/reference/api/require.html#require-async-%E9%93%BE%E5%BC%8F%E8%B0%83%E7%94%A8
        if (node.callee.type === 'Identifier' && node.callee.name === 'require') {
          if (node.arguments[0] && node.arguments[0].type === 'Literal' && typeof node.arguments[0].value === 'string') {
            requireModules.push({
              // @ts-ignore
              start: node.arguments[0].start,
              // @ts-ignore
              end: node.arguments[0].end,
              value: node.arguments[0].value,
            })
          }
        }
        else if (
          node.callee.type === 'MemberExpression'
          && node.callee.object.type === 'Identifier'
          && node.callee.object.name === 'require'
          && node.callee.property.type === 'Identifier'
          && node.callee.property.name === 'async') {
          if (node.arguments[0] && node.arguments[0].type === 'Literal' && typeof node.arguments[0].value === 'string') {
            requireModules.push({
              // @ts-ignore
              start: node.arguments[0].start,
              // @ts-ignore
              end: node.arguments[0].end,
              value: node.arguments[0].value,
              async: true,
            })
          }
        }
      }
    },
  })

  return {
    requireModules,
  }
}
