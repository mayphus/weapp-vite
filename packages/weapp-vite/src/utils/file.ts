import { configExtensions, jsExtensions, templateExtensions, vueExtensions } from '@/constants'
import fs from 'fs-extra'
import path from 'pathe'

export function isJsOrTs(name?: string) {
  if (typeof name === 'string') {
    return jsExtensions.some(x => name.endsWith(`.${x}`))
  }
  return false
}

export function isTemplateRequest(request: string) {
  return request.endsWith('.wxml') || request.endsWith('.html')
}

export function changeFileExtension(filePath: string, extension: string) {
  if (typeof filePath !== 'string') {
    throw new TypeError(`Expected \`filePath\` to be a string, got \`${typeof filePath}\`.`)
  }

  if (typeof extension !== 'string') {
    throw new TypeError(`Expected \`extension\` to be a string, got \`${typeof extension}\`.`)
  }

  if (filePath === '') {
    return ''
  }

  extension = extension ? (extension.startsWith('.') ? extension : `.${extension}`) : ''

  const basename = path.basename(filePath, path.extname(filePath))
  return path.join(path.dirname(filePath), basename + extension)
}

export async function findVueEntry(filepath: string) {
  for (const ext of vueExtensions) {
    const p = changeFileExtension(filepath, ext)
    if (await fs.exists(p)) {
      return p
    }
  }
}

export async function findJsEntry(filepath: string) {
  for (const ext of jsExtensions) {
    const p = changeFileExtension(filepath, ext)
    if (await fs.exists(p)) {
      return p
    }
  }
}

export async function findJsonEntry(filepath: string) {
  for (const ext of configExtensions) {
    const p = changeFileExtension(filepath, ext)
    if (await fs.exists(p)) {
      return p
    }
  }
}

export async function findTemplateEntry(filepath: string) {
  for (const ext of templateExtensions) {
    const p = changeFileExtension(filepath, ext)
    if (await fs.exists(p)) {
      return p
    }
  }
}

export function isTemplate(filepath: string) {
  return templateExtensions.some(ext => filepath.endsWith(`.${ext}`))
}

export function touch(filename: string) {
  const time = new Date()

  try {
    fs.utimesSync(filename, time, time)
  }
  catch {
    fs.closeSync(fs.openSync(filename, 'w'))
  }
}
