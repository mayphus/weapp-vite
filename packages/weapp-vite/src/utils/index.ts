export * from './file'
export * from './json'
export * from './projectConfig'
export * from './scan'
// export * from './wxml'
// export * from './wxs'
