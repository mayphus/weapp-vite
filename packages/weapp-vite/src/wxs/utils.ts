import { addExtension, removeExtensionDeep } from '@weapp-core/shared'

export function normalizeWxsFilename(value: string) {
  return addExtension(removeExtensionDeep(value), '.wxs')
}
