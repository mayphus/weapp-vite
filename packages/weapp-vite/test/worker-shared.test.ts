import CI from 'ci-info'
import fs from 'fs-extra'
import path from 'pathe'
import { createCompilerContext } from '@/createContext'
import logger from '@/logger'
import { getFixture, scanFiles } from './utils'

vi.mock('@/logger', () => ({
  // ...await importOriginal<typeof import('@/logger')>(),
  default: {
    warn: vi.fn(),
    success: vi.fn(),
    info: vi.fn(),
  },
  // warn: vi.fn(),
}))

describe.skipIf(CI.isCI)('worker-shared', () => {
  const cwd = getFixture('worker-shared')
  const distDir = path.resolve(cwd, 'dist')
  beforeEach(async () => {
    await fs.remove(distDir)
    const ctx = await createCompilerContext({
      cwd,
      inlineConfig: {
        build: {
          minify: false,
        },
      },
    })
    await ctx.buildService.build()
  })

  it('dist', async () => {
    expect(await fs.exists(path.resolve(distDir))).toBe(true)

    const files = await scanFiles(distDir)
    expect(files).toMatchSnapshot()
    for (const file of files) {
      const content = await fs.readFile(path.resolve(distDir, file), 'utf-8')
      expect(content).toMatchSnapshot(file)
    }
    expect(logger.success).toHaveBeenCalled()
  })
})
