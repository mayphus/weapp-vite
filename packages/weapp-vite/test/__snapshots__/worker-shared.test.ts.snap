// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`worker-shared > dist > app.js 1`] = `
"App({
  globalData: {},
  onLaunch() {
  }
});
"
`;

exports[`worker-shared > dist > app.json 1`] = `
"{
  "pages": [
    "pages/index/index",
    "pages/index/vue"
  ],
  "window": {},
  "style": "v2",
  "componentFramework": "glass-easel",
  "sitemapLocation": "sitemap.json",
  "workers": {
    "path": "workers",
    "isSubpackage": true
  }
}"
`;

exports[`worker-shared > dist > app.wxss 1`] = `
".app {
  color: red;
}/*$vite$:1*/"
`;

exports[`worker-shared > dist > components/Navbar/Navbar.js 1`] = `
"Component({});
"
`;

exports[`worker-shared > dist > components/Navbar/Navbar.json 1`] = `
"{
  "component": true,
  "styleIsolation": "apply-shared",
  "usingComponents": {}
}"
`;

exports[`worker-shared > dist > components/Navbar/Navbar.wxml 1`] = `
"<view>Navbar</view>
"
`;

exports[`worker-shared > dist > components/Navbar/Navbar.wxss 1`] = `
".navbar {
  color: red;
}/*$vite$:1*/"
`;

exports[`worker-shared > dist > pages/index/bbc.wxs 1`] = `
"module.exports = {
  a: '1',
  v: '2'
};"
`;

exports[`worker-shared > dist > pages/index/cjs.wxs 1`] = `
"function tapName(event, ownerInstance) {
  console.log('tap 撒的撒打算', JSON.stringify(event));
}
module.exports = {
  tapName: tapName
};"
`;

exports[`worker-shared > dist > pages/index/dep.wxs 1`] = `""`;

exports[`worker-shared > dist > pages/index/dep1.wxs 1`] = `""`;

exports[`worker-shared > dist > pages/index/dep2.wxs 1`] = `""`;

exports[`worker-shared > dist > pages/index/esm.wxs 1`] = `
"var x = require('./cjs.wxs');
var xx = './dep.wxs';
var y = "./dep1.wxs";
var z = "./dep2" + '.wxs';
require(xx);
require(y);
require(z);
function tapName(event, ownerInstance) {
  console.log('tap 撒的撒打算', JSON.stringify(event));
}"
`;

exports[`worker-shared > dist > pages/index/index.js 1`] = `
"Page({});
"
`;

exports[`worker-shared > dist > pages/index/index.json 1`] = `
"{
  "usingComponents": {
    "Navbar": "/components/Navbar/Navbar"
  },
  "navigationBarTitleText": "初始模板"
}"
`;

exports[`worker-shared > dist > pages/index/index.wxml 1`] = `
"<view>
  <Navbar />
  <view>Hello World From Weapp-vite!</view>
  <wxs module="xx" src="./bbc.wxs"></wxs>
  <wxs module="xxx" src="./esm.wxs"></wxs>
  <wxs module="xxxx" src="./index.wxs"></wxs>
</view>
"
`;

exports[`worker-shared > dist > pages/index/index.wxs 1`] = `
"module.exports.bar = bar;
module.exports.foo = void 0;
require("./xx.wxs");
var foo = module.exports.foo = '\\'hello world\\' from 我爱着';
function bar(d) {
  return d;
}"
`;

exports[`worker-shared > dist > pages/index/index.wxss 1`] = `
".page {
  color: red;
}/*$vite$:1*/"
`;

exports[`worker-shared > dist > pages/index/xx.wxs 1`] = `""`;

exports[`worker-shared > dist > sitemap.json 1`] = `
"{
  "desc": "关于本文件的更多信息，请参考文档 https://developers.weixin.qq.com/miniprogram/dev/framework/sitemap.html",
  "rules": [
    {
      "action": "allow",
      "page": "*"
    }
  ]
}"
`;

exports[`worker-shared > dist > theme.json 1`] = `
"{
  "light": {},
  "dark": {}
}"
`;

exports[`worker-shared > dist > workers/hello.js 1`] = `
"import { h as hello } from "./util-fxlilyst.js";
worker.onMessage((res) => {
  console.log(res, hello);
});
"
`;

exports[`worker-shared > dist > workers/index.js 1`] = `
"var __getOwnPropNames = Object.getOwnPropertyNames;
var __commonJS = (cb, mod) => function __require() {
  return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
};
import { h as hello } from "./util-fxlilyst.js";
var require_workers = __commonJS({
  "workers/index.js"(exports, module) {
    var module = { exports: {} };
    module.exports = {
      a: 1
    };
    const __CJS__export_default__ = (module.exports == null ? {} : module.exports).default || module.exports;
    const __CJS__import__0__ = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
      __proto__: null,
      default: __CJS__export_default__
    }, Symbol.toStringTag, { value: "Module" }));
    const utils = __CJS__export_default__ || __CJS__import__0__;
    worker.onMessage((res) => {
      console.log(res, utils, hello);
    });
  }
});
export default require_workers();
"
`;

exports[`worker-shared > dist > workers/util-fxlilyst.js 1`] = `
"function hello() {
  return "world";
}
export {
  hello as h
};
"
`;

exports[`worker-shared > dist 1`] = `
[
  "app.js",
  "app.json",
  "app.wxss",
  "components/Navbar/Navbar.js",
  "components/Navbar/Navbar.json",
  "components/Navbar/Navbar.wxml",
  "components/Navbar/Navbar.wxss",
  "pages/index/bbc.wxs",
  "pages/index/cjs.wxs",
  "pages/index/dep.wxs",
  "pages/index/dep1.wxs",
  "pages/index/dep2.wxs",
  "pages/index/esm.wxs",
  "pages/index/index.js",
  "pages/index/index.json",
  "pages/index/index.wxml",
  "pages/index/index.wxs",
  "pages/index/index.wxss",
  "pages/index/xx.wxs",
  "sitemap.json",
  "theme.json",
  "workers/hello.js",
  "workers/index.js",
  "workers/util-fxlilyst.js",
]
`;
